lightning/src/util/persist.rs

```diff
@@ -249,10 +259,1010 @@ where
+		let maybe_old_monitor = self.read_monitor(&monitor_name);
+		match maybe_old_monitor {
+			Ok((_, ref old_monitor)) => {
+				// Check that this key isn't already storing a monitor with a higher update_id (collision)
+				if old_monitor.get_latest_update_id() > monitor.get_latest_update_id() {
+					log_error!(
+						self.logger,
+						"Tried to write a monitor at the same outpoint {} with a higher update_id!",
+						monitor_name.as_str()
+					);
+					return chain::ChannelMonitorUpdateStatus::UnrecoverableError;
+				}
+			}
+			// This means the channel monitor is new.
+			Err(ref e) if e.kind() == io::ErrorKind::NotFound => {}
+			_ => return chain::ChannelMonitorUpdateStatus::UnrecoverableError,
+		}
+		// Serialize and write the new monitor
+		let mut monitor_bytes = Vec::with_capacity(
+			MONITOR_UPDATING_PERSISTER_PREPEND_SENTINEL.len() + monitor.serialized_length(),
+		);
+		monitor_bytes.extend_from_slice(MONITOR_UPDATING_PERSISTER_PREPEND_SENTINEL);
+		monitor.write(&mut monitor_bytes).unwrap();
+		match self.kv_store.write(
+			CHANNEL_MONITOR_PERSISTENCE_NAMESPACE,
+			CHANNEL_MONITOR_PERSISTENCE_SUB_NAMESPACE,
+			monitor_name.as_str(),
+			&monitor_bytes,
+		) {
+			Ok(_) => {
+				// Assess cleanup. Typically, we'll clean up only between the last two known full monitors.
+				if let Ok((_, old_monitor)) = maybe_old_monitor {
+					// Updates must be > the old monitor's update_id. If we can't add, the channel
+					// is already stored closed so we can skip this.
+					if let Some(start) = old_monitor.get_latest_update_id().checked_add(1) {
+						let end = if monitor.get_latest_update_id() == CLOSED_CHANNEL_UPDATE_ID {
+							// We don't want to clean the rest of u64, so just do possible pending updates. Note that
+							// we never write updates at `CLOSED_CHANNEL_UPDATE_ID`.
+							cmp::min(
+								start.saturating_add(self.maximum_pending_updates),
+								CLOSED_CHANNEL_UPDATE_ID - 1,
+							)
+						} else {
+							monitor.get_latest_update_id().saturating_sub(1)
+						};
+						// We should bother cleaning up only if there's at least one update expected.
+						if end >= start {
+							for update_id in start..=end {
+								let update_name = UpdateName::from(update_id);
+								#[cfg(debug_assertions)]
+								{
+									// For closed channels, we unavoidably issue some no-op deletes.
+									if monitor.get_latest_update_id() != CLOSED_CHANNEL_UPDATE_ID {
+										debug_assert_eq!(
+											self.read_monitor_update(&monitor_name, &update_name)
+												.unwrap_or_else(|_| panic!(
+													"failed to read monitor update {}",
+													update_name.as_str()
+												))
+												.update_id,
+											update_name.0
+										);
+									}
+								}
+								if let Err(e) = self.kv_store.remove(
+									CHANNEL_MONITOR_UPDATE_PERSISTENCE_NAMESPACE,
+									monitor_name.as_str(),
+									update_name.as_str(),
+									true,
+								) {
+									log_error!(
```

Comment:
The argument for it is that a wrapper class around MUP would be able to get the error and handle it appropriately (as it sits, they'll never know). The argument against it is that an out-of-the-box user would get an UnrecoverableError here that isn't actually unrecoverable at all. I think of the three who have commented on this (besides me, who had it as a TODO), two have suggested it should _not_ error. 

It's hard to imagine how this would fail (especially a lazy delete) a few clock cycles after a write succeeded, isn't it? If it did, wouldn't that mean the next write is likely to fail, which we're designed to yield an error on?

Reply:
I think that's good guidance. Since we can't expect a lazy delete to _ever_ occur, we can only issue them in places where they are non-critical, by definition. Arguably, we should have made it an infallible alternative method instead of a flag, but I think this is reasonably easy to remember / reason your way into.